src/OrbitCaptureClient/ApiEventProcessorTest.cpp

```diff
@@ -0,0 +1,238 @@
+  void OnThreadStateSlice(orbit_client_protos::ThreadStateSliceInfo) override {}
+  void OnAddressInfo(orbit_client_protos::LinuxAddressInfo) override {}
+  void OnUniqueTracepointInfo(uint64_t, orbit_grpc_protos::TracepointInfo) override {}
+  void OnTracepointEvent(orbit_client_protos::TracepointEventInfo) override {}
+};
+
+// Test CaptureListener used to validate TimerInfo data produced by api events.
+class ApiEventCaptureListener : public EmptyCaptureListener {
+ public:
+  void OnTimer(const orbit_client_protos::TimerInfo& timer_info) override {
+    timers_.emplace_back(timer_info);
+  }
+  std::vector<TimerInfo> timers_;
+};
+
+// ApiTester exposes methods that mocks the Orbit API and internally creates grpc events that it
+// forwards to both a CaptureEventProcessor and a ApiEventProcessor. The test makes sure that both
+// processors forward the same information to their listener. Note that a CaptureEventProcessor owns
+// a ApiEventProcessor to which it forwards ApiEvent events. To help readability of test code, the
+// api methods return a ApiTester reference so that we can chain function calls.
+class ApiTester {
+ public:
+  ApiTester()
+      : api_event_processor_(&api_event_listener_),
+        capture_event_processor_(&capture_event_listener_) {}
+
+  ApiTester& Start(const char* name = nullptr, orbit_api_color color = kOrbitColorAuto) {
+    EnqueueApiEvent(orbit_api::kScopeStart, name, 0, color);
+    return *this;
+  }
+  ApiTester& Stop() {
+    EnqueueApiEvent(orbit_api::kScopeStop);
+    return *this;
+  }
+  ApiTester& StartAsync(uint64_t id, const char* name = nullptr,
+                        orbit_api_color color = kOrbitColorAuto) {
+    EnqueueApiEvent(orbit_api::kScopeStartAsync, name, id, color);
+    return *this;
+  }
+  ApiTester& StopAsync(uint64_t id) {
+    EnqueueApiEvent(orbit_api::kScopeStopAsync, /*name=*/nullptr, id);
+    return *this;
+  }
+  ApiTester& TrackValue(const char* name, uint64_t value, orbit_api_color color = kOrbitColorAuto) {
+    EnqueueApiEvent(orbit_api::kTrackUint64, name, value, color);
+    return *this;
+  }
+
+  ApiTester& CheckNumTimers(size_t num_timers) {
+    EXPECT_EQ(api_event_listener_.timers_.size(), num_timers);
+    EXPECT_EQ(capture_event_listener_.timers_.size(), num_timers);
+    return *this;
+  }
+
+  size_t CountTimersOfType(orbit_api::EventType type) {
+    const std::vector<TimerInfo>& timers = api_event_listener_.timers_;
+    return std::count_if(timers.begin(), timers.end(), [type](const TimerInfo& timer_info) {
+      return ApiEventFromTimerInfo(timer_info).type == type;
+    });
+  }
+
+  ApiTester& CheckNumScopeTimers(size_t num_timers) {
+    EXPECT_EQ(CountTimersOfType(orbit_api::kScopeStart), num_timers);
+    return *this;
+  }
+
+  ApiTester& CheckNumAsyncScopeTimers(size_t num_timers) {
+    EXPECT_EQ(CountTimersOfType(orbit_api::kScopeStartAsync), num_timers);
+    return *this;
+  }
+
+  ApiTester& CheckNumTrackingTimers(size_t num_timers) {
```

Comment:
I would suggest renaming these methods from `Check...` to `Expect...`. They better hint at the fact that `EXPECT_...` is used and *not* `CHECK`, which has a different meaning.

Reply:
Done, thanks.